lib/src/iterable_matchers.dart

```diff
@@ -165,30 +165,39 @@ class _UnorderedMatches extends Matcher {
         return 'has too many elements (${list.length} > ${_expected.length})';
       }
 
-      var matched = new List<bool>.filled(list.length, false);
-      var expectedPosition = 0;
-      for (var expectedMatcher in _expected) {
-        var actualPosition = 0;
-        var gotMatch = false;
-        for (var actualElement in list) {
-          if (!matched[actualPosition]) {
-            if (expectedMatcher.matches(actualElement, {})) {
-              matched[actualPosition] = gotMatch = true;
-              break;
-            }
+      var adjacency = list
+          .map((_) => new List.filled(_expected.length, false, growable: false))
```

Comment:
Done. In the small value case it probably makes memory usage a bit worse since I suspect the default empty list allocates more memory than these fixed lists - but I think it make the code more clear and should be an improvement for large (sparsely matching) inputs.